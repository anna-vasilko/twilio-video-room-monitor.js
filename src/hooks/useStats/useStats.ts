import React from 'react';
import { StatsReport } from 'twilio-video';
import { RoomStatsContext } from '../../components/RoomStatsProvider/RoomStatsProvider';

export default function useStats() {
  const stats = React.useContext(RoomStatsContext);
  return stats;
}

function getAllTracks(statsReport: StatsReport) {
  const { localAudioTrackStats, localVideoTrackStats, remoteAudioTrackStats, remoteVideoTrackStats } = statsReport;

  const allTracks = [
    ...localAudioTrackStats,
    ...localVideoTrackStats,
    ...remoteAudioTrackStats,
    ...remoteVideoTrackStats,
  ];

  return allTracks;
}

export function getTrackData(trackSid: string, statsReport: StatsReport) {
  const allCurrentTracks = getAllTracks(statsReport);
  return allCurrentTracks.filter((t) => t.trackSid === trackSid);
}

const sum = (arr: number[]) => arr.reduce((a, b) => a + b, 0);
const round = (num: number) => Math.round((num + Number.EPSILON) * 10) / 10;

export function useTrackBandwidth(trackSid: string) {
  const { stats, previousStats } = useStats();
  if (!stats || !previousStats) return null;

  const currentTrackData = getTrackData(trackSid, stats.statsReports[0]);
  const previousTrackData = getTrackData(trackSid, previousStats.statsReports[0]);

  if (
    currentTrackData.length === 0 ||
    previousTrackData.length === 0 ||
    currentTrackData.length !== previousTrackData.length
  ) {
    return null;
  }

  const currentBytes = sum(currentTrackData.map((d) => d.bytesReceived ?? d.bytesSent ?? 0));
  const previousBytes = sum(previousTrackData.map((d) => d.bytesReceived ?? d.bytesSent ?? 0));

  const currentTime = currentTrackData[0]?.timestamp;
  const previousTime = previousTrackData[0]?.timestamp;

  return round((currentBytes - previousBytes) / (currentTime / previousTime) / 1000);
}

export function useTotalBandwidth(kind: 'bytesSent' | 'bytesReceived') {
  const { stats, previousStats } = useStats();
  if (!stats || !previousStats) return null;

  const currentTrackData = getAllTracks(stats.statsReports[0]);
  const previousTrackData = getAllTracks(previousStats.statsReports[0]);

  if (currentTrackData.length === 0 || previousTrackData.length === 0) return null;

  const currentBytes = sum(currentTrackData.map((d) => d[kind] ?? 0));
  const previousBytes = sum(previousTrackData.map((d) => d[kind] ?? 0));

  const currentTime = currentTrackData[0]?.timestamp;
  const previousTime = previousTrackData[0]?.timestamp;

  return round((currentBytes - previousBytes) / (currentTime / previousTime) / 1000);
}
