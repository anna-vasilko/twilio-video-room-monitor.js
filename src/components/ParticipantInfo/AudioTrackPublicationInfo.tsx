import React from 'react';
import styled from 'styled-components';
import {
  LocalAudioTrack,
  LocalAudioTrackPublication,
  RemoteAudioTrack,
  RemoteAudioTrackPublication,
} from 'twilio-video';
import useIsTrackEnabled from '../../hooks/useIsTrackEnabled/useIsTrackEnabled';
import { useTrackBandwidth } from '../../hooks/useStats/useStats';
import useTrack from '../../hooks/useTrack/useTrack';
import { theme } from '../theme';
import Datum from '../typography/Datum/Datum';

const AudioTrackInfo: React.FC<{
  track: LocalAudioTrack | RemoteAudioTrack;
}> = ({ track }) => {
  const isEnabled = useIsTrackEnabled(track);

  return (
    <>
      <Datum label="isEnabled" value={String(isEnabled)} />
    </>
  );
};

const Container = styled.div`
  &:not(:last-child) {
    border-bottom: 1px solid ${theme.borderColor};
    margin-bottom: 3px;
  }
  padding-bottom: 3px;
`;

export const AudioTrackPublicationInfo: React.FC<{
  publication: LocalAudioTrackPublication | RemoteAudioTrackPublication;
}> = ({ publication }) => {
  const track = useTrack(publication) as LocalAudioTrack | RemoteAudioTrack | undefined;
  const trackBandwidth = useTrackBandwidth(publication.trackSid);

  return (
    <Container>
      <Datum label="Name" value={publication.trackName} />
      <Datum label="SID" value={publication.trackSid} />
      <Datum label="isSubscribed" value={String(!!track)} />
      <Datum label="bandwidth" value={String(trackBandwidth) + 'kbps'} />
      {track && <AudioTrackInfo track={track} />}
    </Container>
  );
};
