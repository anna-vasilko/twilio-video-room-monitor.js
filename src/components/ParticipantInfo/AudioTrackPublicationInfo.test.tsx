import React from 'react';
import { shallow } from 'enzyme';
import { AudioTrackPublicationInfo, AudioTrackInfo } from './AudioTrackPublicationInfo';
import { useTrackBandwidth, useTrackData } from '../../hooks/useStats/useStatsUtils';
import useTrack from '../../hooks/useTrack/useTrack';

jest.mock('../../hooks/useIsTrackEnabled/useIsTrackEnabled', () => () => true);
jest.mock('../../hooks/useStats/useStatsUtils');
jest.mock('../../hooks/useTrack/useTrack');

const mockUseTrackBandwidth = useTrackBandwidth as jest.Mock<any>;
const mockUseTrackData = useTrackData as jest.Mock<any>;
const mockUseTrack = useTrack as jest.Mock<any>;

mockUseTrackBandwidth.mockImplementation(() => 1234.56);
mockUseTrackData.mockImplementation(() => ({ codec: 'testCodec', jitter: 2, packetsLost: 7 }));
mockUseTrack.mockImplementation(() => 'testTrack');

describe('the AudioTrackInfo component', () => {
  it('should render correctly if an audio track is present', () => {
    const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly if there is no audio track data', () => {
    mockUseTrackData.mockImplementationOnce(() => null);
    const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper.find({ label: 'Codec' }).exists()).toBeFalsy();
  });

  it('should render correctly for the RemoteAudioTrack when packetsLost is not defined', () => {
    mockUseTrackData.mockImplementationOnce(() => (
      { codec: 'testCodec', frameRate: null, packetsLost: null}
    ));
    const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly for the RemoteAudioTrack when both packetsReceived and packetsLost are defined', () => {
    mockUseTrackData.mockImplementationOnce(() => (
      { codec: 'testCodec', frameRate: null, packetsLost: 7, packetsReceived: 100 }
    ));
    const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly for the LocalVideoTrack when both packetsSent and packetsLost are defined', () => {
    mockUseTrackData.mockImplementationOnce(() => (
      { codec: 'testCodec', frameRate: null, packetsLost: 56, packetsSent: 183 }
    ));
    const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly for the RemoteVideoTrack when packetsLost is 0 and packetsReceived is defined', () => {
    mockUseTrackData.mockImplementationOnce(() => (
      { codec: 'testCodec', frameRate: null, packetsLost: 0, packetsReceived: 10 }
    ));
    const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly for the LocalVideoTrack when packetsLost is 0 and packetsSent is defined', () => {
    mockUseTrackData.mockImplementationOnce(() => (
      { codec: 'testCodec', frameRate: null, packetsLost: 0, packetsSent: 10 }
    ));
    const wrapper = shallow(<AudioTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('the AudioTrackPublicationInfo component', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <AudioTrackPublicationInfo publication={{ trackName: 'testName', trackSid: 'testSid' } as any} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render the AudioTrackInfo component if an audio track is not present', () => {
    mockUseTrack.mockImplementationOnce(() => null);
    const wrapper = shallow(
      <AudioTrackPublicationInfo publication={{ trackName: 'testName', trackSid: 'testSid' } as any} />
    );
    expect(wrapper.find(AudioTrackInfo).exists()).toBeFalsy();
  });
});
