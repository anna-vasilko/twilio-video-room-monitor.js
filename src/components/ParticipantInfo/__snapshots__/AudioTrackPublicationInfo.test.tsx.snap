// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the AudioTrackInfo component should render correctly for the LocalVideoTrack when both packetsSent and packetsLost are defined 1`] = `
<Fragment>
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Jitter"
    value="undefined"
  />
  <Memo(Datum)
    label="Packets Lost"
    value="56"
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value="30.601%"
  />
</Fragment>
`;

exports[`the AudioTrackInfo component should render correctly for the LocalVideoTrack when packetsLost is 0 and packetsSent is defined 1`] = `
<Fragment>
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Jitter"
    value="undefined"
  />
  <Memo(Datum)
    label="Packets Lost"
    value="0"
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value="0%"
  />
</Fragment>
`;

exports[`the AudioTrackInfo component should render correctly for the RemoteAudioTrack when both packetsReceived and packetsLost are defined 1`] = `
<Fragment>
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Jitter"
    value="undefined"
  />
  <Memo(Datum)
    label="Packets Lost"
    value="7"
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value="7%"
  />
</Fragment>
`;

exports[`the AudioTrackInfo component should render correctly for the RemoteAudioTrack when packetsLost is not defined 1`] = `
<Fragment>
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Jitter"
    value="undefined"
  />
  <Memo(Datum)
    label="Packets Lost"
    value="null"
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value="undefined%"
  />
</Fragment>
`;

exports[`the AudioTrackInfo component should render correctly for the RemoteVideoTrack when packetsLost is 0 and packetsReceived is defined 1`] = `
<Fragment>
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Jitter"
    value="undefined"
  />
  <Memo(Datum)
    label="Packets Lost"
    value="0"
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value="0%"
  />
</Fragment>
`;

exports[`the AudioTrackInfo component should render correctly if an audio track is present 1`] = `
<Fragment>
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Jitter"
    value="2"
  />
  <Memo(Datum)
    label="Packets Lost"
    value="7"
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value="undefined%"
  />
</Fragment>
`;

exports[`the AudioTrackPublicationInfo component should render correctly 1`] = `
<styled.div>
  <Memo(Datum)
    label="Name"
    value="testName"
  />
  <Memo(Datum)
    label="SID"
    value="testSid"
  />
  <Memo(Datum)
    label="isSubscribed"
    value="true"
  />
  <AudioTrackInfo
    track="testTrack"
    trackSid="testSid"
  />
</styled.div>
`;
